### ISSUE TITLE
Develop mechanism to replace sleeping

### ISSUE THREAD

**Issue by nealef** [anup-kodlekere/gaplib#2]
The current build scripts use arbitrary sleep times to allow things to settle before proceeding.

**Comment by anup-kodlekere** [comment]
@nealef what do you think of this?

```bash

max_time=60  

start_time=$(date +%s)
end_time=$((start_time + timeout))

while true; do
    current_time=$(date +%s)

    if lxc info "${BUILD_CONTAINER}" | yq '.Status'; then
        echo "${BUILD_CONTAINER} is up and running."
        break
    fi

    if ((current_time >= end_time)); then
        echo "Container boot-up timeout reached."
        exit 1
    fi

    echo "Waiting for container ${BUILD_CONTAINER} to start..."
    sleep 2
done
```

**Comment by anup-kodlekere** [comment]
I haven't fully tested it. I'll need to figure out a way to mock a container boot-up failure or something along those lines.

**Comment by nealef** [comment]
Rather than a start/end time just use a retry count. Because you are sleeping 2 seconds each time you can determine the retry count via timeout/2. Saves playing around with date output.

**Comment by anup-kodlekere** [comment]
Yeah, makes sense. Initially I did play around with a counter based mechanism but figured it was too "simple".

**Comment by nealef** [comment]
Don't underestimate simple! ;-)

**Comment by anup-kodlekere** [comment]
```bash

#!/bin/bash

retry_count=0
timeout=10
sleep_period=2
max_retries=$((timeout/sleep_period))

while [ $retry_count -lt $max_retries ]; do
    if [ "$(lxc info "${BUILD_CONTAINER}" | yq '.Status')" = "RUNNING" ]; then
        echo "${BUILD_CONTAINER} is up and running."
        break
    else
        echo "Attempt $((retry_count+1)): Waiting for ${BUILD_CONTAINER} to start."
        ((retry_count++))
        sleep $sleep_period
    fi
done

if [ $retry_count -eq $max_retries ]; then
    echo "Container boot-up timeout reached. Exiting."
fi
```

**Comment by anup-kodlekere** [comment]
I do have a question though, if we follow the same approach as you've drawn-up for Alma (but with Ubuntu), would that make things better?

**Comment by nealef** [comment]
> I do have a question though, if we follow the same approach as you've drawn-up for Alma (but with Ubuntu), would that make things better?

What in particular? I use Alma to avoid having to build the dotnet SDK which is an asynchronous process to the runner build meaning the two have to be coordinated. By using the RPM I avoid that step. Ubuntu don't ship a `.deb` for .NET so we cannot do the same thing.

**Comment by nealef** [comment]
Suggestion...
```
timeout=10
sleep_period=2
max_retries=$((timeout/sleep_period))
for (( retry_count = 0; retry_count < max_retries; retry_count++ ))
do
    if [ "$(lxc info "${BUILD_CONTAINER}" | yq '.Status')" = "RUNNING" ]; then
        echo "${BUILD_CONTAINER} is up and running."
        break
    else
        echo "Attempt ${retry_count}: Waiting for ${BUILD_CONTAINER} to start."
        sleep ${sleep_period}
    fi
done

if [ ${retry_count} -eq ${max_retries} ]; then
    echo "Container boot-up timeout reached. Exiting." >&2
    # Perform any cleanup 
    exit 1
fi

**Comment by anup-kodlekere** [comment]
> What in particular?

Using the distrobuilder approach to build the LXD image for Power.

> Ubuntu don't ship a `.deb` for .NET so we cannot do the same thing.

Ahh yes, I keep forgetting that for some reason.

**Comment by anup-kodlekere** [comment]
Fixed in #4
